### PR TITLE
Update debugger network code

### PR DESCRIPTION
**Pull request by MisterDA**
The OCaml debugger sets up a socket between the debugger and the runtime of the process being debugged. The user has some control over the socket: she can specify an IPv4 address or host, or a path to a Unix domain socket. They are selected with `ocamldebug -s [address]` option, or interactively with `set socket [address]`, and with the `CAML_DEBUG_SOCKET` environment variable to the remotely debugged process.

With some warnings enabled, C compilers now complain that some network functions are deprecated (e.g., `gethostbyname` is replaced by `getaddrinfo`), because the functions don't support IPv6.

This PR:

- replaces uses of `gethostbyname` with `getaddrinfo` in the runtime and in `ocamldebug`;
- allows using IPv6 addresses, specified using the bracketed syntax `[::1]:8080`, and IPv6 hosts;
- allows using Unix domain socket on Windows;
- modernizes and cleans up the network code.

~My code tries to detect whether a colon `:` separating an IP address and a port has been specified; if not it considers that the user has specified an Unix domain socket (a path in the filesystem). Then, if the user has specified an implicit path or something that doesn't look like an IPv6 address, my code also picks an Unix domain socket. All the remaining cases (IPv4 address, IPv6 address, hostname; followed by a port) are handled by `getaddrinfo` after some little processing.~

I've tested this code on macOS and Windows, with IPv4, IPv6 addresses and hosts, and absolute and relative paths for Unix domain sockets.

### CODE REVIEW

**Comment by gasche** [comment]
Both `[foo]bar` and `foo:bar` are valid paths for files or directories on my Linux system, which means that some valid paths cannot be used for local sockets.

This was already an existing issue with the previous code (I think?) which used the "does it contain a `:`?" criterion.

I think that the following approaches would fix it, if we decide that it is worth fixing:
- have a convention to clarify that something really is a path name, for example starting with `/` or `./` (for absolute and relative paths)
- enrich the UI to specify the communication channel with an explicit type, for example `--socket-path` and `--socket-addr` as refinements of the `-s` option, etc.

**Comment by MisterDA** [comment]
> Both `[foo]bar` and `foo:bar` are valid paths for files or directories on my Linux system

I probably wasn't clear enough in the PR description (and I discovered [`Filename.is_implicit`](https://v2.ocaml.org/api/Filename.html#VALis_implicit) just after submitting). On my macOS, `[foo]bar` is correctly identified as a file path. `foo:bar` is identified as an address, with `foo` being the host, and `bar` the protocol name associated to a port number that `getaddrinfo` should retrieve, if I'm not mistaken.

I think I see the problem this way: almost anything can be a path, but there are things that are clearly not pairs of (host, port), if they contain directory separators and square brackets. I think it's better to deal with the obvious cases, let `getaddrinfo` sort the rest; and the user can always specify a relative path `./` to get out of this mess. We should never identify a valid `host:port` as a path, that would be a bug. Then there's the problem of error reporting; if I wrote `localhost:8080x` I probably wasn't trying to create a domain socket.

> have a convention to clarify that something really is a path name, for example starting with `/` or `./` (for absolute and relative paths)

Yes. Non-implicit paths are well recognized as such.

> Return `true` if the file name is relative and does not start with an explicit reference to the current directory (`./` or `../` in Unix), `false` if it starts with an explicit reference to the root directory or the current directory.

~With this code, I consider that the address looks like a file path if:~
- ~it doesn't contain `:`, or~
- ~it is not implicit and it does not look like an IPv6 address (enclosed in square brackets).~

Possible refinements, but a bit more annoying to implement, include checking whether the address contains a directory separator, and whether the segment after the last colon parses as an integer.

> enrich the UI to specify the communication channel with an explicit type, for example `--socket-path` and `--socket-addr` as refinements of the `-s` option, etc.

That would be the clearest option, but what should we do about the `-s` option? And we would also need to give refinements to the `CAML_DEBUG_SOCKET` environment variable on the other side.

**Comment by MisterDA** [comment]
My main goal was to update the network code. I found out while hacking that I could add IPv6 and Unix domain sockets support too (on Windows). If this feature is too complicated to sort out, I could remove it (the last commit!) from this PR and re-submit later.

File: debugger/unix_tools.ml [hunk 1]
```diff
@@ -22,24 +22,24 @@ open Unix
...
-      "Win32" -> failwith "Unix sockets not supported"
-    | _ -> (PF_UNIX, ADDR_UNIX address)
+  let is_ipv6 str n = str.[0] = '[' && str.[n] = ']' in
+  let un_addr =
+    { ai_family = PF_UNIX; ai_socktype = SOCK_STREAM; ai_protocol = 0;
+      ai_addr = ADDR_UNIX address; ai_canonname = ""; } in
+  match String.rindex address ':' with
+  | exception Not_found -> un_addr
+  | n ->
+     let is_ipv6 = is_ipv6 address (n - 1) in
```
Comment (ghost) [inline debugger/unix_tools.ml]: There is no guarantee that `n - 1` is `>= 0` here (and same with the `n - 2` computation further down).

Comment (ghost) [inline debugger/unix_tools.ml]: The use of the same identifier for both a function and its result is confusing here IMHO.

File: runtime/debugger.c [hunk 2]
```diff
@@ -170,16 +169,38 @@ static void winsock_cleanup(void)
...
+{
+  size_t len = strlen(path);
+  return (len < 1 || path[0] != '/')
+    && strncmp(path, "./", 2) != 0
+    && strncmp(path, "../", 3) != 0;
+}
 #endif
+static bool is_ipv6(const char *address, const char *port)
+{
+  return address[0] == '[' && port[-1] == ']';
```
Comment (ghost) [inline runtime/debugger.c]: I am not really fond of `port[-1]` here - if you pass `":"` as the debugger socket specification, you will invoke this function with `port == address` and only skirt the out-of-bounds access because it will fail the `address[0] == '['` check.

Comment (MisterDA) [reply runtime/debugger.c]: > only skirt the out-of-bounds access because it will fail the `address[0] == '['` check.

I was proud of this trick but I understand if you think that's not a great style.

Comment (MisterDA) [reply runtime/debugger.c]: I'm now checking that there's a least 4 bytes in the address (`[::]`) before accessing both `address` and `port`.

File: debugger/unix_tools.ml (hunk 1 above)
Comment (MisterDA) [reply debugger/unix_tools.ml]: I'm now checking that there's a least 4 bytes in the address (`[::]`) before accessing it.

Comment (MisterDA) [reply debugger/unix_tools.ml]: I've changed to `is_likely_ipv6` and `likely_ipv6`.

File: runtime/debugger.c [hunk 3]
```diff
@@ -170,16 +169,40 @@ static void winsock_cleanup(void)
...
   WSACleanup();
 }
+
+/* from Filename.is_implicit */
+static bool is_implicit(const char *path)
+{
+  size_t len = strlen(path);
+  return (len < 1 || path[0] != '/')
+    && (len < 1 || path[0] != '\\')
+    && (len < 1 || path[1] == 0 || path[1] != ':')
```
Comment (ghost) [inline runtime/debugger.c]:
```suggestion
    && (len < 2 || path[1] == 0 || path[1] != ':')
```

Comment (MisterDA) [reply runtime/debugger.c]: Argh, thanks.

**Comment by gasche** [comment]
My general impression:

- this looks reasonable overall (my question about path vs. address confusion was a nitpick)
- it was not a wise choice to couple a refactoring/cleanup with new features ; next time I would recommend sending separate PRs
- nothing in the PR is critical correctness-wise, and I think it is beneficial to share knowledge of this codebase
- all this considered, I am in favor of merging the PR once someone is done looking at the details, and it looks like @dustanddreams is doing this work

**Comment by ghost** [comment]
>     * all this considered, I am in favor of merging the PR once someone is done looking at the details, and it looks like @dustanddreams is doing this work

But you have no proof that I am not @MisterDA's sock puppet!

**Review by ghost — APPROVED** [review]

File: debugger/unix_tools.ml [hunk 4]
```diff
@@ -22,24 +22,27 @@ open Unix
...
-                 failwith "Can't convert address"),
-            (try int_of_string port with Failure _ ->
-               prerr_endline "The port number should be an integer";
-               failwith "Can't convert address")))
-  with Not_found ->
-    match Sys.os_type with
-      "Win32" -> failwith "Unix sockets not supported"
-    | _ -> (PF_UNIX, ADDR_UNIX address)
+  if String.length address == 0 then
+    failwith "Can't convert address: empty or invalid address";
```
Comment (gasche) [inline debugger/unix_tools.ml]: It's `= 0`, not `== 0`. Also the error message is weird because here we know for sure that the address is empty, not invalid, right?

File: debugger/unix_tools.ml [hunk 5]
```diff
@@ -22,24 +22,27 @@ open Unix
...
+  let is_likely_ipv6 str n =
+    n >= 4 && str.[0] = '[' && str.[n] = ']' in
+  let un_addr =
```
Comment (gasche) [inline debugger/unix_tools.ml]: For me the name `un_addr` is not clear, I think this value means "a unix socket rather than a network socket", or "a path rather than an address" and the name does not reflect it, making reading the logic below harder than it should be.

File: debugger/unix_tools.ml [hunk 6]
```diff
@@ -22,24 +22,27 @@ open Unix
...
+    { ai_family = PF_UNIX; ai_socktype = SOCK_STREAM; ai_protocol = 0;
+      ai_addr = ADDR_UNIX address; ai_canonname = ""; } in
+  match String.rindex address ':' with
+  | exception Not_found -> un_addr
+  | n ->
+     let likely_ipv6 = is_likely_ipv6 address (n - 1) in
+     if not (Filename.is_implicit address) && not likely_ipv6 then
```
Comment (gasche) [inline debugger/unix_tools.ml]: I find the logic in this block of code too hard to follow. Before your change the idea was rather clear: either there is a `:` and we handle this as an IP (V4) address, or there isn't and we assume a unix socket path. Now what is the logic, and is it apparent in the code?

For example I think that it is not clear what `not (Filename.is_implicit ...)` is testing for, and maybe this could be helped with variable naming: `let is_an_explicit_path = not (Filename.is_implicit address) in ...`.

**Review by gasche — COMMENTED** [review]
I find the logic to select the socket type not clear enough. The code should be clearer. Maybe we also want a comment that explains what the logic is (it is not documented anywhere so far?), but this might not be necessary if the code is readable enough as its own specification.

File: debugger/unix_tools.ml (hunk 4 above)
Comment (MisterDA) [reply debugger/unix_tools.ml]: Right. I was wondering what to do if the user had specified a string containing nul bytes, but if they do that they probably have other problems.

File: debugger/unix_tools.ml (hunk 6 above)
Comment (MisterDA) [reply debugger/unix_tools.ml]: Well a path like `C:\TMP\my.sock` or `./my:sock` is obviously not a `host:port`. `not (Filename.is_implicit ...)` caches paths that are rooted or (syntactically) relative to the current directory (they begin with `./` or `../`).
If we only look at whether the address contains a `:` to decide whether it's an Unix domain socket or an inet address, then we'll miss absolute paths on Windows.
I'll look for a simpler compromise.

File: debugger/unix_tools.ml (hunk 5 above)
Comment (MisterDA) [reply debugger/unix_tools.ml]: Thanks. I'll see if I can eliminate the multiple uses of this variable, or name it `unix_addr_info` which I hope conveys the right meaning.

File: debugger/unix_tools.ml (hunk 6 above)
Comment (gasche) [reply debugger/unix_tools.ml]: I'm not saying that the previous logic was better -- I agree it was worse, especially on Windows. I am asking for a clear description of what the new logic is, and code that matches this clear description.

**Comment by MisterDA** [comment]
> I find the logic to select the socket type not clear enough. The code should be clearer.

I hope to have made it simpler and clearer.

> Maybe we also want a comment that explains what the logic is (it is not documented anywhere so far?), but this might not be necessary if the code is readable enough as its own specification.

I hope that the code is self-sufficient, although it does not justify the choices made. I now consider that the string is a Unix domain socket path if:
- if it does not contain a `:` (can't be a `host:port` pair);
- it starts with an explicit reference to a root directory or the current directory (it's an explicit path).
I then let `getaddrinfo` sort out the rest after splitting `host` and `port` strings. I also try to have both the OCaml and C code behave identically.

Sorry for the back and forth, I tried to be too clever without achieving a satisfying result. Are the code and the explanations satisfying now? Do you think that the code covers reasonable use-cases and has sane failure conditions?

File: debugger/unix_tools.ml [hunk 7]
```diff
@@ -22,24 +22,28 @@ open Unix
...
+  | exception Not_found -> unix_addr_info
+  | _ when not (Filename.is_implicit address) -> unix_addr_info
+  | n ->
+     let is_likely_ipv6 =
+       n >= 4 && address.[0] = '[' && address.[n - 1] = ']' in
+     let host = if is_likely_ipv6 then String.sub address 1 (n - 2)
+                else String.sub address 0 n
+     and port = String.(sub address (n + 1) (length address - n - 1)) in
+     if host = "" || port = "" then
+       failwith "Can't convert address: empty host or empty port";
```
Comment (gasche) [inline debugger/unix_tools.ml]: Nit (here and below): I think that it is useful to print the invalid address, so I would use `Can't convert address %S: ..` (the %S format quotes the string).

File: debugger/unix_tools.ml [hunk 8]
```diff
@@ -22,24 +22,28 @@ open Unix
...
-      "Win32" -> failwith "Unix sockets not supported"
-    | _ -> (PF_UNIX, ADDR_UNIX address)
+  if address = "" then
+    failwith "Can't convert address: empty address";
+  let unix_addr_info =
+    { ai_family = PF_UNIX; ai_socktype = SOCK_STREAM; ai_protocol = 0;
+      ai_addr = ADDR_UNIX address; ai_canonname = ""; } in
+  match String.rindex address ':' with
+  | exception Not_found -> unix_addr_info
+  | _ when not (Filename.is_implicit address) -> unix_addr_info
```
Comment (gasche) [inline debugger/unix_tools.ml:33]: I think that short comment saying that, say `(* "./foo" is explicitly a path and not a network address *)` could help understand what Filename.is_implicit does here, it is not obvious.

**Review by gasche — APPROVED** [review]
Yes, I think this is reasonable now. Thanks! (See minor comments.)

File: debugger/unix_tools.ml (hunk 7 above)
Comment (MisterDA) [reply debugger/unix_tools.ml]: Done, thanks.

File: debugger/unix_tools.ml (hunk 8 above)
Comment (MisterDA) [reply debugger/unix_tools.ml:33]: Indeed… I've added your comment.
